sycl/include/CL/sycl/detail/spirv.hpp

```diff
@@ -290,6 +290,258 @@ AtomicMax(multi_ptr<T, AddressSpace> MPtr, intel::memory_scope Scope,
+}
+
+// Bitcast shuffles can be implemented using a single SPIR-V SubgroupShuffle
+// intrinsic, but require type-punning via an appropriate integer type
+template <typename T>
+using EnableIfBitcastShuffle =
+    detail::enable_if_t<!detail::is_arithmetic<T>::value &&
+                            (std::is_trivially_copyable<T>::value &&
+                             (sizeof(T) == 1 || sizeof(T) == 2 ||
+                              sizeof(T) == 4 || sizeof(T) == 8)),
+                        T>;
+
+template <typename T>
+using ConvertToNativeShuffleType_t = select_cl_scalar_integral_unsigned_t<T>;
+
+template <typename T>
+EnableIfBitcastShuffle<T> SubgroupShuffle(T x, id<1> local_id) {
+  using ShuffleT = ConvertToNativeShuffleType_t<T>;
+  auto ShuffleX = detail::bit_cast<ShuffleT>(x);
+  ShuffleT Result = __spirv_SubgroupShuffleINTEL(
+      ShuffleX, static_cast<uint32_t>(local_id.get(0)));
+  return detail::bit_cast<T>(Result);
+}
+
+template <typename T>
+EnableIfBitcastShuffle<T> SubgroupShuffleXor(T x, id<1> local_id) {
+  using ShuffleT = ConvertToNativeShuffleType_t<T>;
+  auto ShuffleX = detail::bit_cast<ShuffleT>(x);
+  ShuffleT Result = __spirv_SubgroupShuffleXorINTEL(
+      ShuffleX, static_cast<uint32_t>(local_id.get(0)));
+  return detail::bit_cast<T>(Result);
+}
+
+template <typename T>
+EnableIfBitcastShuffle<T> SubgroupShuffleDown(T x, T y, id<1> local_id) {
+  using ShuffleT = ConvertToNativeShuffleType_t<T>;
+  auto ShuffleX = detail::bit_cast<ShuffleT>(x);
+  auto ShuffleY = detail::bit_cast<ShuffleT>(y);
+  ShuffleT Result = __spirv_SubgroupShuffleDownINTEL(
+      ShuffleX, ShuffleY, static_cast<uint32_t>(local_id.get(0)));
+  return detail::bit_cast<T>(Result);
+}
+
+template <typename T>
+EnableIfBitcastShuffle<T> SubgroupShuffleUp(T x, T y, id<1> local_id) {
+  using ShuffleT = ConvertToNativeShuffleType_t<T>;
+  auto ShuffleX = detail::bit_cast<ShuffleT>(x);
+  auto ShuffleY = detail::bit_cast<ShuffleT>(y);
+  ShuffleT Result = __spirv_SubgroupShuffleUpINTEL(
+      ShuffleX, ShuffleY, static_cast<uint32_t>(local_id.get(0)));
+  return detail::bit_cast<T>(Result);
+}
+
+// Generic shuffles may require multiple calls to SPIR-V SubgroupShuffle
+// intrinsics, and should use the fewest shuffles possible
+template <typename T>
+using EnableIfGenericShuffle =
+    detail::enable_if_t<!detail::is_arithmetic<T>::value &&
+                            !(std::is_trivially_copyable<T>::value &&
+                              (sizeof(T) == 1 || sizeof(T) == 2 ||
+                               sizeof(T) == 4 || sizeof(T) == 8)),
+                        T>;
+
+template <typename T>
+EnableIfGenericShuffle<T> SubgroupShuffle(T x, id<1> local_id) {
+  T Result;
+  char *XBytes = reinterpret_cast<char *>(&x);
+  char *ResultBytes = reinterpret_cast<char *>(&Result);
+  int Offset = 0;
+#pragma unroll
+  for (; Offset < sizeof(T); Offset += sizeof(uint64_t)) {
```

Comment:
You're right that the first iteration of the loop is going to do an extra `memcpy` there, and we should be guarding it with a branch.  I'll fix that.

I think we should keep the logic as looking only at `sizeof(T)` rather than involving the number of bytes remaining, though:
1) I believe the 8 byte copy branch is the only one that needs a loop.  Multiple 4 byte copies can be handled by the 8 byte branch, multiple 2 byte copies can be handled by the 4 byte branch and so on.
2) The compiler isn't smart enough to optimize away the branches if they depend on `Remaining`: https://godbolt.org/z/KiBDRV

I'll see if I can simplify the logic to make this clearer.

Reply:
I've simplified this in 75d1d4c and 1db9b27.  It looks quite a bit nicer now and is hopefully easier to review.